querier/src/ingester/circuit_breaker.rs

```diff
@@ -0,0 +1,1258 @@
+        let not_cancelled = test_signal.unwrap_or_else(|| Arc::new(AtomicBool::new(true)));
+        let fut = TrackedFuture::new(fut, not_cancelled);
+        let res = fut.await;
+
+        let is_error = if let Err(e) = &res {
+            match e {
+                FlightClientError::Flight {
+                    source: _source @ FlightError::GrpcError(e),
+                } => !matches!(
+                    e.code(),
+                    tonic::Code::NotFound | tonic::Code::ResourceExhausted
+                ),
+                FlightClientError::Connecting { .. }
+                | FlightClientError::Handshake { .. }
+                | FlightClientError::Flight { .. } => true,
+                // do NOT break circuit for client-side errors
+                FlightClientError::CreatingRequest { .. } => false,
+                FlightClientError::CircuitBroken { .. } => {
+                    unreachable!("circuit breaker wrapped into circuit breaker")
+                }
+            }
+        } else {
+            false
+        };
+
+        if is_error {
+            let mut circuits = self.circuits.lock();
+
+            let now = self.time_provider.now();
+
+            match circuits.entry(Arc::clone(&ingester_addr)) {
+                Entry::Vacant(_) => unreachable!("should have been inserted already"),
+                Entry::Occupied(mut o) => {
+                    let o = o.get_mut();
+                    let maybe_backoff_and_metrics = match o {
+                        Circuit::Open { gen, .. } => {
+                            assert_ne!(
+                                start_gen, *gen,
+                                "could not have started in an open circuit state"
+                            );
+                            None
+                        }
+                        Circuit::HalfOpen {
+                            backoff,
+                            metrics,
+                            gen,
+                            ..
+                        } => {
+                            assert_eq!(
+                                start_gen, *gen,
+                                "there's only a single concurrent request for half-open circuits"
+                            );
+                            Some((backoff.take().expect("not moved"), metrics.clone()))
+                        }
+                        Circuit::Closed {
+                            metrics,
+                            error_count,
+                            gen,
+                        } => {
+                            if *gen == start_gen {
+                                *error_count += 1;
+                                (*error_count >= self.open_circuit_after_n_errors).then(|| {
+                                    warn!(
+                                        ingester_address = ingester_addr.as_ref(),
+                                        "Circuit opened"
```

Comment:
```suggestion
                                        "Error contacting ingester, circuit opened"
```

It might also be helpful to log the actual error (e.g. res)